cli/pkg/helm/client.go

```diff
@@ -0,0 +1,566 @@
+					Getters:          c.Providers,
+					RepositoryConfig: c.Settings.RepositoryConfig,
+					RepositoryCache:  c.Settings.RepositoryCache,
+					Out:              os.Stdout,
+				}
+				if err := man.Update(); err != nil {
+					return err
+				}
+
+				// as chart dependencies are updated fetch the chart again
+				details.helmChart, *details.chartPath, err = c.getChart(details.spec.ChartName, details.chartPathOptions)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// upgrade upgrades a chart and CRDs
+func (c *HelmClient) upgrade(ctx context.Context, spec *ChartSpec) error {
+	client := action.NewUpgrade(c.ActionConfig)
+	mergeUpgradeOptions(spec, client)
+
+	if client.Version == "" {
+		client.Version = ">0.0.0-0"
+	}
+
+	helmChart, chartPath, err := c.getChart(spec.ChartName, &client.ChartPathOptions)
+	if err != nil {
+		return err
+	}
+
+	d := &updateDeps{
+		helmChart:        helmChart,
+		spec:             spec,
+		chartPath:        &chartPath,
+		chartPathOptions: &client.ChartPathOptions,
+	}
+	err = c.updateDependencies(d)
+	if err != nil {
+		return fmt.Errorf("failed to update helm chart dependencies %v", err)
+	}
+
+	values, err := spec.GetValuesMap()
+	if err != nil {
+		return err
+	}
+
+	if c.linting {
+		err = c.lint(chartPath, values)
+		if err != nil {
+			return err
+		}
+	}
+
+	if !spec.SkipCRDs && spec.UpgradeCRDs {
+		log.Printf("updating crds")
+		err = c.upgradeCRDs(ctx, helmChart)
+		if err != nil {
+			return err
+		}
+	}
+
+	rel, err := client.Run(spec.ReleaseName, helmChart, values)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("release upgrade successfully: %s/%s-%s", rel.Name, rel.Name, rel.Chart.Metadata.Version)
+
+	return nil
+}
+
+// uninstallRelease uninstalls the provided release
+func (c *HelmClient) uninstallRelease(spec *ChartSpec) error {
+	client := action.NewUninstall(c.ActionConfig)
+
+	mergeUninstallReleaseOptions(spec, client)
+
+	resp, err := client.Run(spec.ReleaseName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("release removed, response: %v", resp)
+
+	return nil
+}
+
+// lint lints a chart's values
+func (c *HelmClient) lint(chartPath string, values map[string]interface{}) error {
+	client := action.NewLint()
+
+	result := client.Run([]string{chartPath}, values)
+
+	for _, err := range result.Errors {
+		log.Printf("Error %s", err)
+	}
+
+	if len(result.Errors) > 0 {
+		return fmt.Errorf("linting for chartpath %q failed", chartPath)
+	}
+
+	return nil
+}
+
+// upgradeCRDs upgrades the CRDs of the provided chart
+func (c *HelmClient) upgradeCRDs(ctx context.Context, chartInstance *chart.Chart) error {
+	cfg, err := c.Settings.RESTClientGetter().ToRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	k8sClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	for _, crd := range chartInstance.CRDObjects() {
+		// use this ugly detour to parse the crdYaml to a CustomResourceDefinitions-Object because direct
```

Comment:
did you get this logic from somewhere?

Reply:
This to update the CRDs & this logic borrowed from the community helm-client.

Reply:
could you add a comment with the link to where the code is borrowed from?